### ISSUE TITLE
Streaming

### ISSUE THREAD

**Issue by cyrilchapon** [sequelize/sequelize#10347]
- [This issue](https://github.com/sequelize/sequelize/issues/2454) is 3 years old
- it does seem like a major feature
- you asked for community contribution, several time
- community answered and proposed help and contribution several times
- no-one answered appart the annoying bot

?

Related https://github.com/sequelize/sequelize/issues/4286 (which has been closed as duplicate without taking time on the main issue)

**Comment by knoxcard** [comment]
Seems super useful and fast!

**Comment by cyrilchapon** [comment]
yeah but actualy, node.js stream is just like an "implementation" of a well-known-and-empiric feature of database drivers : **cursors**

- https://www.postgresql.org/docs/9.2/plpgsql-cursors.html
- https://dev.mysql.com/doc/refman/8.0/en/cursors.html
- https://mariadb.com/kb/en/library/cursor-overview/
- https://docs.microsoft.com/en-us/sql/ado/guide/data/understanding-cursors-and-locks?view=sql-server-2017

I'd try to answer here the question asked #2454 by @ericnini
> It may be that an ORM is not the right tool for this job to begin with.

## Philosophy :
Yes it does. Or at least it's not exclusive.
ORM stands for *O*bject *R*elational *M*apper. It "maps" *objects* with *relational data*.
Nothing in that title argues for "ORM = you have an big array when you SELECT data".

In my opinion, an ORM does 3 main things :
- When you have a *row* coming from a db, it transforms it, objectify it and type checks and casts it before returning it to you
- It helps you query the database in a more language-compliant way
- It abstracts relations (and foreign keys) to object links

Nothing of that is incompatible with a cursor streamed strategy. (at least for reading data)

Cursor is just a "versus array buffer" reading strategy, and actualy not having cursor is quite an enormous limitation. It basically ties the query to the available-amount-of-ram in which you can in-memory store the result of the query.

## Use case

None of the "massive query" use-cases I can think of are incompatible with the concept of an ORM.
A few examples :
- A migration script between 2 applications. You select 2M rows in the database, and push it to an API on the other application. It takes 2 hours, requires some handling between both. For example the source data has some Guid and Dates typed fields that you don't want to cast yourself. This is one of the points of an ORM : formating data to objects
- A "download an export" API endpoint. You select 2M rows in the database, and download it through a json / CSV file. Both [json](https://www.npmjs.com/package/stream-json) and [CSV](https://csv.js.org/) are streamable, both to disk file or HTTP response.
- A job that's scheduled every night that fetches and update data in a whole database, syncing with another service.

**Yes**, for each of these use-cases, one can argue that "an ORM is not the right tool for this". My question is simply *WHY ?*. Why on earth, massive querying, parsing, and mapping shouldn't benefit of an ORM powers ? This does seem like a Symfony and Doctrine answer to me, and is a too-simple answer to a real need. Even with massive querying strategy, one **does** need type checking, columns parsing, etc. It also needs query-helping provided when you have an ORM (For example, the power of abstracting computed arguments querying)

## Examples

Mongoose does it pretty well. There are several projects where I use both Mongo and SQL databases, and I like to think Mongoose like the Sequelize for MongoDB. The API is not ideal, but the result is perfect. Of course, MongoDB is not multi-database, thus the bottom-up code strategy to be able to query both to a whole array or to expose a streamed cursor is simpler. But still it does the job.

## API

Well, the API in Mongoose is pretty simple. Why not inspiring from them ?

The choice between "a whole" or "a cursor" is simply "attaching a .then" or "calling .cursor"

```
MyModel.find({})
  .then(modelArray => {})
```

OR

```
const readStream = MyModel.find({}).stream()
```

**Comment by knoxcard** [comment]
Now I can't sleep!  I need this now.

**Comment by sushantdhiman** [comment]
@cyrilchapon Thanks for your comments, I think API has been discussed before and it was concluded that an option `stream: true` works best for Sequelize case.

```js
Model.findAll({ stream: true }).then((stream) => { });
```

Next phase will be implementation, which someone in need of this feature will have to take care of.

@knoxcard We don't ban for silly reasons, We all want better tools for our jobs. It is natural to compare various tools and learn from them

**Comment by cyrilchapon** [comment]
Hey @sushantdhiman, and thanks for answering. This API seems very OK too.

I think now everyone here is ok to say streaming (wired to cursor) is a must, looking indeed at knex, mongoose and probably others ORM / CRM.

Next step as you say is implementation. I agree. @mellogarrett asked for core contributors help in https://github.com/sequelize/sequelize/issues/2454 months ago, without answer. I can help too if pointed in the right direction. Please, let's organize it and get shit done ?

**Comment by NicolasCaous** [comment]
I'm dying for this feature. :cry: Is someone implementing it?

**Comment by NicolasCaous** [comment]
Don't know if it is a dumb thing to do, but would this feature also include streams for large BLOBs?

Let's say we have a row with a blob column containing 5MB of data. I wanted to stream it to better performance. Should this become a new issue?

**Comment by cyrilchapon** [comment]
@NicolasCaous I think so. This is pretty different as it does not involve cursor

**Comment by papb** [comment]
@NicolasCaous I think that deserves a separate issue indeed, please open one and link it here.

@cyrilchapon Thanks for all the input so far :grimacing: I see you (and others) have been asking about "a direction", can you be more specific? If there is anything I can answer to help you get started it would be great. But I am pretty sure you understand the topic better than I do :+1:

By the way, I have one suggestion to begin: [your comment above](https://github.com/sequelize/sequelize/issues/10347#issuecomment-454358357) was great IMO, I would like to ask you to propose an specific API that you would like most, and show some snippet examples of how those would be used. Not only for `find`, but also `create`, `update`, and such. Thanks!!

**Comment by papb** [comment]
Also, I have another question, does [contra's comment on the other issue](https://github.com/sequelize/sequelize/issues/2454#issuecomment-470669708) work for you?

**Comment by contra** [comment]
@papb FYI - that code is pretty old, the most recent version is here: https://github.com/staeco/iris-ql/blob/master/src/util/export.js (iris-ql is a layer on top of sequelize that supports streaming via this function)

**Comment by papb** [comment]
@contra Excellent, thank you!

Yes I agree, the API should be agreed first. I asked @cyrilchapon to suggest one (for `find`, `create`, `update`, etc), if you could suggest one as well that would be great!

**Comment by contra** [comment]
Find All:

```js
const stream = Model.findAll({ stream: true })
```

Find And Count All:

```js
const stream = Model.findAndCountAll({ stream: true })
stream.on('count', (count) => {
  // count emitted separately from standard data events
  // up to you how you want to use it
})
```

Create:

```js
newModels.pipe(Model.create({ stream: true }))
```

Create:

```js
newModels.pipe(Model.create({ stream: true }))
```
Update:

The use case for this would just be updating records by ID? IMO this could be left out.

```js
modelUpdates.pipe(Model.update({ stream: true }))
```

**Comment by papb** [comment]
@contra Thanks! Can you add the preparation and usage parts as well, i.e., a simple way to use the `stream` and how `newModels` / `modelUpdates` were defined?

**Comment by contra** [comment]
@papb Updated

**Comment by cyrilchapon** [comment]
## API

@papb I don't have much time to expand on every questions over there (sorry for that).
I'll focus on the blocking one :

@papb
> I would like to ask you to propose an specific API that you would like most, and show some snippet examples of how those would be used. Not only for find, but also create, update, and such.

@sushantdhiman
> @cyrilchapon Thanks for your comments, I think API has been discussed before and it was concluded that an option stream: true works best for Sequelize case.
> `Model.findAll({` stream: true }).then((stream) => { });`

@contra
> Find All:
> const stream = Model.findAll({ stream: true })
> stream.on('data', (inst) => {
>  // inst = a model instance
> })

There is a slight difference between @sushantdhiman 's and @contra 's APIs : the Promise versus the sync stream gather.

I'd like to highlight that iris-ql from @contra is only working on PG atm, and for this to work `const stream = Model.findAll({ stream: true })` (synchronous); we have to **make sure** that every sequelize-compliant dialects (and underlying connectors) are able to get the cursor stream synchroneously. Otherwise, we'll have to make it asynchronous `const stream = await Model.findAll({ stream: true })` (notice `await`). That's a first and important decision.

Apart from stream instanciation, every single point of @contra proposal seems very OK for me.

## Implementation

@papb 
> I see you (and others) have been asking about "a direction", can you be more specific?

Well, sequelize is a pretty large codebase. For example, for the `find` query, I'd need to know :
- Where would be the best place to "parse `stream` option, and decide the underlying query format / underlying connector call" (where are actual underlying connector calls, actually)
- Where could be the best place to "bypass Promise-based return value"
- What is the best function(s) to call on an actual returned database row, in order to transform it to a Model instance ?

This is all for the `find` I guess. The last one seems to be answered by iris-ql (thanks for this).

**Comment by cyrilchapon** [comment]
Trying to answer to myself

> we have to make sure that every sequelize-compliant dialects (and underlying connectors) are able to get the cursor stream synchroneously

Sequelize's documentation says
> ```
> $ npm install --save sequelize # This will install v5
> 
> # And one of the following:
> $ npm install --save pg pg-hstore # Postgres
> $ npm install --save mysql2
> $ npm install --save mariadb
> $ npm install --save sqlite3
> $ npm install --save tedious # Microsoft SQL Server
> ```

I assume `pg`, `mysql2`, `mariadb`, `sqlite3`, `tedious` are the right places to search for that.

⚠️This is all about **reading** data

### Postgres
Apparently, [`pg`](https://github.com/brianc/node-postgres) with a plugin [`pg-query-stream`](https://github.com/brianc/node-pg-query-stream) (through [`pg-cursor`](https://github.com/brianc/node-pg-cursor)).

Stream instanciation is **synchronous**

> ```
>   const query = new QueryStream('SELECT * FROM generate_series(0, $1) num', [1000000])
>   const stream = client.query(query)
> ```

The plugin (and it's underlying dependency) is writen by `pg` author, @brianc. I assume it to be stable.

### Mysql

[`mysql2`](https://github.com/sidorares/node-mysql2) is supposed to be a [`mysql`](https://github.com/mysqljs/mysql) [drop-in-place API replacement](https://github.com/sidorares/node-mysql2#api-and-configuration), thus supposed to implement [this strategy](https://github.com/mysqljs/mysql#streaming-query-rows).

Stream instanciation is **synchronous**
> ```
> var query = connection.query('SELECT * FROM posts');
> query
>   .on('error', function(err) {
>     // Handle error, an 'end' event will be emitted after this as well
>   })
>   .on('fields', function(fields) {
>     // the field packets for the rows to follow
>   })
>   .on('result', function(row) {
>     // Pausing the connnection is useful if your processing involves I/O
>     connection.pause();
> 
>     processRow(row, function() {
>       connection.resume();
>     });
>   })
>   .on('end', function() {
>     // all rows have been received
>   });
> ```

I assume this is ok, because of reading the [source code here](https://github.com/sidorares/node-mysql2/blob/58c1056edd21ad42e2e755959b75f76b502e3f72/lib/results_stream.js#L7-L37)

### MariaDb

[`mariadb`](https://github.com/MariaDB/mariadb-connector-nodejs) seems to [have this feature at its core](https://github.com/MariaDB/mariadb-connector-nodejs/blob/master/documentation/promise-api.md#connectionquerystreamsql-values--emitter). 

Stream instanciation is **synchronous**

> ```
> connection.queryStream("SELECT * FROM mysql.user")
>       .on("error", err => {
>         console.log(err); //if error
>       })
>       .on("fields", meta => {
>         console.log(meta); // [ ...]
>       })
>       .on("data", row => {
>         console.log(row);
>       })
>       .on("end", () => {
>         //ended
>       });
> ```

### SQLite

[`sqlite3`](https://github.com/mapbox/node-sqlite3) doesn't seem to have cursor stream at its core (issue [here](https://github.com/mapbox/node-sqlite3/issues/686)); however it has an asynchronous method [`.each(query)`](https://github.com/mapbox/node-sqlite3/wiki/API#databaseeachsql-param--callback-complete) bound to an underlying db cursor, which should be easy to wrap inside a `ReadableStream`. An [attempt here in `streamsql`](https://github.com/brianloveswords/streamsql/blob/105ab77fea7b0aca043ef4ec19bb3dd5712bc037/lib/drivers/sqlite3.js#L14-L32)

Wrapping this `.each()` inside a `ReadableStream` would be possible **synchronously**

```
  db.each("SELECT rowid AS id, info FROM lorem", function(err, row) {
      console.log(row.id + ": " + row.info);
  });
```

### tedious

[`tedious`](http://tediousjs.github.io/tedious) exposes a `Request` object which has a pseudo-stream interface ([`on('row')` event](http://tediousjs.github.io/tedious/api-request.html#event_row)), bound to an underlying database cursor. This should be easy to wrap inside a ReadableStream.

This would be **synchronous** (as creating the Request is synchronous)

```
var Request = require('tedious').Request;

request = new Request("select 42, 'hello world'", function(err, rowCount) {
  ...
});
connection.execSql(request);
request.on('row', function (columns) { });
```

(there is a callback on the constructor signature (wtf), but the instanciation is synchronous)

(Also, there seem to be a [Stream with generators interface](https://github.com/tediousjs/tedious/pull/285), but I don't know where)

**Comment by cyrilchapon** [comment]
For insertion :

- PG has [`pg-copy-stream`](https://github.com/brianc/node-pg-copy-streams) which uses a `COPY TO` query with a readable stream... but this sounds more like a "hack", doesn't it ?
- Mariadb [has this feature at its core](https://www.npmjs.com/package/mariadb#insert-streaming)
- Mysql [has this feature at its core](https://github.com/sidorares/node-mysql2/blob/master/documentation/Extras.md#sending-tabular-data-with-load-infile-and-local-stream)
- SQLite doesn't have it [though, `streamsql` gave it a try 5 years ago...](https://github.com/brianloveswords/streamsql/blob/105ab77fea7b0aca043ef4ec19bb3dd5712bc037/index.js#L338-L417)
- Tedious [surprisingly has this at its core!](https://tediousjs.github.io/tedious/bulk-load.html#function_getRowStream)

**Comment by contra** [comment]
@cyrilchapon I have to disagree with your assessment that my stream example is synchronous - what happens under the hood of the stream can be whatever you want. The PG example implemented in iris-ql is not synchronous, it needs to grab a connection and execute a query to get a cursor before it can start emitting data events. If anything fails while grabbing a connection or getting the cursor, an error event is emitted. A stream isn't really isn't much different than a promise, except you get multiple resolutions over time (data events) instead of a single resolved value.

**Comment by cyrilchapon** [comment]
@contra, I never said your stream is synchronous. I said your stream **creation** is synchronous.

**Comment by contra** [comment]
@cyrilchapon Creation of both promises and streams is synchronous so it isn't really relevant to the conversation IMO. Streams and promises are both asynchronous resolvers at their core, construction of either is "synchronous" so having a promise that returns a stream is superfluous. Just trying to clarify here because your statements suggested that returning a stream somehow limits it to only postgres, which is not true.

**Comment by jsanta** [comment]
Made a gist with Postgres pg-query-stream and Express:
https://gist.github.com/jsanta/d386a3f908243070c3a6ec9f2a75531b

It requires the SQL query string as input, but you can already generate it using Sequelize's QueryGenerator.
Hope it helps.

**Comment by pittersnider** [comment]
Hi, is this still valid?

**Comment by imjordanxd** [comment]
Has there been any progress on this?

**Comment by rbrcurtis** [comment]
Here's how I solved this with `pg-cursor`: https://gist.github.com/rbrcurtis/e27cec8e924b9adaa7d0340e20cdb999

**Comment by bradisbell** [comment]
If there is no traction on implementing this feature, are there any thoughts for implementing along the side in a best practice way?  That is, is there any way to reuse the Sequelize query builder, but then execute the query ourselves on the underlying connection using the specific methods for whichever connection adapter we're using?

**Comment by contra** [comment]
@bradisbell If you look earlier in this thread I posted a fully working example (and linked to a module you can use if you don't want the code in your own repo), we've been using it in production for almost 3 years without issue.

**Comment by hp8wvvvgnj6asjm7** [comment]
https://github.com/evansiroky/db-streamer

**Comment by ShivamJoker** [comment]
Wow I guess the maintainers really don't want this feature. They want us to move to another library.

**Comment by razarsharp** [comment]
I really need this feature, but don't know how i can push this topic forward. What can we do to make this happen?

**Comment by ShivamJoker** [comment]
You can try knex.js it has streaming support with typescript as well.

**Comment by dudek-igor** [comment]
Still open 😭

**Comment by ephys** [comment]
Purely because having different return types based on parameters is a typing nightmare, I think we should go with the [initial suggestion of `Model.findAllStream()`](https://github.com/sequelize/sequelize/issues/2454#issuecomment-60128989)

But nevertheless, I personally don't have the bandwidth to take on such a large feature right now. If someone wants to open a PR to implement streaming, it's welcomed (we can start with just `findAllStream`)

**Comment by colinhowe** [comment]
Would the maintainers of this repo accept a PR that implements the proposed `findAllStream`?

- What if that PR only supported one dialect with the right structure to  make adding more dialects easy?
- Would there be any objection to making it return a generator or similar?

**Comment by fzn0x** [comment]
> Would the maintainers of this repo accept a PR that implements the proposed `findAllStream`?
> 
> * What if that PR only supported one dialect with the right structure to  make adding more dialects easy?
> * Would there be any objection to making it return a generator or similar?

Yes, it would be great.

- We can start with MySQL
- I would prefer something like this https://github.com/sequelize/sequelize/issues/2454#issuecomment-60128989 so we don't really break the current APIs and the typings is not hard.

What do you think? @WikiRik @sdepold

**Comment by ephys** [comment]
I concur, that would be awesome

> What if that PR only supported one dialect with the right structure to make adding more dialects easy?

That's absolutely fine in my opinion, as long as the other dialects throw an error along the lines of `findAllStream is not yet supported in <dialect>, PR welcome`

**Comment by ephys** [comment]
Forgot to comment on this:

> Would there be any objection to making it return a generator or similar?

That is another feature we're considering: `findAllInBatch`, which would return an async iterable (so an async generator would be a good way to implement it)

**Comment by cyrilchapon** [comment]
> The names are a mouthful. Name ideas very welcome

> Purely because having different return types based on parameters is a typing nightmare, I think we should go with the https://github.com/sequelize/sequelize/issues/2454#issuecomment-60128989

Sincerely, I'd challenge this statement in a near-2023 ORM...

**Comment by ephys** [comment]
What do you mean?

**Comment by cyrilchapon** [comment]
@ephys yeah I guess. I mean, they're here for that kind of purpose; the API would be much cleaner.

Typing is indeed a mess with return types and arguments; but frankly that's not impossible and my opinion API cleanness > typing simplicity.

**Comment by ephys** [comment]
Note that my problem with overloads changing the return type isn't with making our typing more complicated on our end, but more complicated for our users.

If TypeScript isn't sure of which overload to use (which easily happens if you're building FindOptions dynamically), the return type will be the union of all possible return types.

Since we want to have first-class support for TypeScript, having different return types based on the inputs is something we'd like to avoid as much as possible.

There is also the issue of the input option bag type: these stateful methods are going to need a batchSize option, the stateless ones are going to need options like `after`, `before`, `first`, and `last`. None of those are relevant to the base findAll

**Comment by mbsanchez01** [comment]
Any news about this feature ?

**Comment by TechnotronicOz** [comment]
Streaming is foundational to so many implementations, almost so much that nearly all other libraries treat it ubiquitously. At this point, it almost feels as if Sequelize is dying given the lack of traction over such a long period for a feature considered to be standard.

* http://c2fo.github.io/patio/patio_Dataset.html#stream
* https://typeorm.io/select-query-builder#streaming-result-data
* https://kysely-org.github.io/kysely-apidoc/interfaces/SelectQueryBuilder.html#stream
* https://knexjs.org/guide/interfaces.html#streams
* https://node-postgres.com/apis/cursor via `pg-cursor`
* https://www.npmjs.com/package/pg-query-stream

**Comment by mihairaulea** [comment]
a. This issue was 1st opened on [Oct 21, 2014](https://github.com/sequelize/sequelize/issues/2454#issue-46336169)
That's 10 years, 1 months and 28 days ago.

b. Meanwhile, the package has been downloaded between 1 MILLION and 2 MILLION times per week. 

I think billing people for executing open source software(in the cents for thousands of executions) would open a talent marketplace big enough to solve issues like this one in 2 days, not 10 years. 
Using a transparent ledger will also open a fair pay marketplace which will incentivize adoption amongst developers.

And we would soon take over software - us, who actually build it :)

I have spent my last 8 years looking for the protocol that can achieve this. I think i am close to wrapping it up. It will have one rule, easy to understand, applied recursively.

[Are you curious enough?](https://openspur.com/)

**Comment by yocontra** [comment]
I think given that there is a viable (and relatively small) workaround (https://github.com/sequelize/sequelize/issues/2454#issuecomment-470669708) I understand why this has been kicked down in priority but would still really like to see this someday! Unfortunately I don't have the time to contribute a PR as I originally offered when I posted my workaround due to a new role but would be glad to help anyone who wants to tackle it.

**Comment by Solicitouscouncilor38** [comment]
@yocontra https://lnkr.fun/bec47ed
